File: examples/android/MSDKSampleAppJava/lib/src/main/java/msdk/java/types/MessageAttachment.java
package msdk.java.types;

import android.util.Base64;

import com.evernym.sdk.vcx.VcxException;
import com.evernym.sdk.vcx.connection.ConnectionApi;
import com.evernym.sdk.vcx.utils.UtilsApi;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.List;
import java.util.Objects;
import java.util.concurrent.ExecutionException;

import java9.util.concurrent.CompletableFuture;
import msdk.java.messages.Message;
import msdk.java.utils.CommonUtils;

public class MessageAttachment {
    public String type;
    public JSONObject data;

    public MessageAttachment(String type, JSONObject data) {
        this.type = type;
        this.data = data;
    }

    public static MessageAttachment parse(String invite) {
        try {
            String attachment = UtilsApi.vcxExtractAttachedMessage(invite).get();
            JSONObject attachmentJson = CommonUtils.convertToJSONObject(attachment);
            assert attachmentJson != null;
            attachmentJson.put("@id", getIdFromInvite(invite));
            String type = attachmentJson.getString("@type");
            return new MessageAttachment(type, attachmentJson);
        } catch (VcxException | JSONException | InterruptedException | ExecutionException e) {
            e.printStackTrace();
        }
        return null;
    }

    private static String getIdFromInvite(String invite) {
        try {
            JSONObject inviteJson = CommonUtils.convertToJSONObject(invite);

            if (inviteJson != null && !inviteJson.has("request~attach")) {
                return null;
            }

            String requestAttachCode = inviteJson.getString("request~attach");
            JSONArray requestsAttachItems = new JSONArray(requestAttachCode);
            if (requestsAttachItems.length() == 0) {
                return null;
            }

            JSONObject requestsAttachItem = requestsAttachItems.getJSONObject(0);
            return requestsAttachItem.getString("@id");
        } catch (JSONException exp) {
            exp.printStackTrace();
        }
        return null;
    }

    public boolean isCredentialAttachment() {
        return this.type.contains("issue-credential");
    }

    public boolean isProofAttachment() {
        return this.type.contains("present-proof");
    }
}

File: examples/android/MSDKSampleAppKotlin/app/src/main/java/msdk/kotlin/sample/types/MessageAttachment.kt
package msdk.kotlin.sample.types

import com.evernym.sdk.vcx.VcxException
import com.evernym.sdk.vcx.utils.UtilsApi
import msdk.kotlin.sample.utils.CommonUtils
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.util.concurrent.ExecutionException

class MessageAttachment(var type: String, var data: JSONObject) {
    val isCredentialAttachment: Boolean
        get() = type.contains("issue-credential")

    val isProofAttachment: Boolean
        get() = type.contains("present-proof")

    companion object {
        fun parse(invite: String?): MessageAttachment? {
            try {
                val attachment = UtilsApi.vcxExtractAttachedMessage(invite).get()
                val attachmentJson = CommonUtils.convertToJSONObject(attachment)!!
                attachmentJson.put("@id", getIdFromInvite(invite))
                val type = attachmentJson.getString("@type")
                return MessageAttachment(type, attachmentJson)
            } catch (e: VcxException) {
                e.printStackTrace()
            } catch (e: JSONException) {
                e.printStackTrace()
            } catch (e: InterruptedException) {
                e.printStackTrace()
            } catch (e: ExecutionException) {
                e.printStackTrace()
            }
            return null
        }

        private fun getIdFromInvite(invite: String?): String? {
            try {
                val inviteJson = CommonUtils.convertToJSONObject(invite)
                if (inviteJson != null && !inviteJson.has("request~attach")) {
                    return null
                }
                val requestAttachCode = inviteJson!!.getString("request~attach")
                val requestsAttachItems = JSONArray(requestAttachCode)
                if (requestsAttachItems.length() == 0) {
                    return null
                }
                val requestsAttachItem = requestsAttachItems.getJSONObject(0)
                return requestsAttachItem.getString("@id")
            } catch (exp: JSONException) {
                exp.printStackTrace()
            }
            return null
        }
    }
}